homeassistant/components/tuya/config_flow.py

```diff
@@ -1,143 +1,216 @@
+    __reauth_entry: ConfigEntry | None = None
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self.__login_control = LoginControl()
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Step user."""
+        errors = {}
+        placeholders = {}
+
+        if user_input is not None:
+            success, response = await self.__async_get_qr_code(
+                user_input[CONF_USER_CODE]
+            )
+            if success:
+                return await self.async_step_scan()
+
+            errors["base"] = "login_error"
+            placeholders = {
+                TUYA_RESPONSE_MSG: response.get(TUYA_RESPONSE_MSG, "Unknown error"),
+                TUYA_RESPONSE_CODE: response.get(TUYA_RESPONSE_CODE, "0"),
+            }
+        else:
+            user_input = {}
+
+        return self.async_show_form(
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USER_CODE, default=user_input.get(CONF_USER_CODE, "")
+                    ): str,
+                }
+            ),
+            errors=errors,
+            description_placeholders=placeholders,
+        )
 
-        for app_type in ("", TUYA_SMART_APP, SMARTLIFE_APP):
-            data[CONF_APP_TYPE] = app_type
-            if data[CONF_APP_TYPE] == "":
-                data[CONF_AUTH_TYPE] = AuthType.CUSTOM
-            else:
-                data[CONF_AUTH_TYPE] = AuthType.SMART_HOME
-
-            api = TuyaOpenAPI(
-                endpoint=data[CONF_ENDPOINT],
-                access_id=data[CONF_ACCESS_ID],
-                access_secret=data[CONF_ACCESS_SECRET],
-                auth_type=data[CONF_AUTH_TYPE],
+    async def async_step_scan(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Step scan."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="scan",
+                description_placeholders={
+                    TUYA_RESPONSE_QR_CODE: self.__qr_image,
+                },
             )
-            api.set_dev_channel("hass")
 
-            response = api.connect(
-                username=data[CONF_USERNAME],
-                password=data[CONF_PASSWORD],
-                country_code=data[CONF_COUNTRY_CODE],
-                schema=data[CONF_APP_TYPE],
+        ret, info = await self.hass.async_add_executor_job(
+            self.__login_control.login_result,
+            self.__qr_code,
+            TUYA_CLIENT_ID,
+            self.__user_code,
+        )
+        if not ret:
+            return self.async_show_form(
+                step_id="scan",
+                errors={"base": "login_error"},
+                description_placeholders={
+                    TUYA_RESPONSE_QR_CODE: self.__qr_image,
+                    TUYA_RESPONSE_MSG: info.get(TUYA_RESPONSE_MSG),
```

Comment:
What will `None` as placeholder value mean?

Reply:
👍  Adjusted it to be similar to the rest of the handling